feat(cli,client,generator-helper,sdk): add ability for generators to print use message

muxahuk opened the pull request:
This MR addresses https://github.com/prisma/prisma/issues/1737 issue

muxahuk committed:
feat(cli,client,generator-helper,sdk): add ability for generators to print use message

implement a way for generators to print use message after it has been successfully generated. This addresses #1737 issue

muxahuk commented on src/packages/cli/src/Generate.ts:
```diff
@@ -74,6 +74,10 @@ export class Generate implements Command {
             `✔ Generated ${chalk.bold(name!)}${version ? ` (version: ${version})` : ''
             }${toStr} in ${formatms(after - before)}\n`,
           )
+          const useMessage = await generator.getUseMessage()
```
this promise never finishes for generators compiled without generator-helper changes...

muxahuk replied:
this is resolved here: GeneratorProcess.ts like 207
```
const timeout = setTimeout(
        () => resolve(''),
        1000,
      )
```

timsuchanek commented:
Thanks for the PR @muxahuk!
What exactly is the breaking change here?
Btw to ensure compatibility: You can check in the SDK, that if it is the `prisma-client-js` provider and its use message is empty, add what we had before. This way we make sure it's still compatible with the old clients, which don't yet send the use message.

muxahuk commented:
in short:
1) we have multiple generators for our schema
2) we update prisma
3) when prisma generate is being run it will send message to generator (running in fork/spawn) with "useMessage" code
4) generator that was generated without this update would never "catch" that message and therefor would never return any response to "main" thread
5) promise that we are awaiting here: https://github.com/prisma/prisma/pull/4231#discussion_r521714218 will never finish leading to "generate" command hanging for ever.

so, i was trying to find some way to get to know what version of generator-helper generator is using to determine whether functionality with "getUseMessage" is supported or not to be able to "fail" fast, but not finding one - i implemented a "setTimeout" workaround to return empty string if generator is compiled with old "generator-helper".

So, before i did that workaround - it was breaking change since in order for generators to work with new version of "prisma generate" they would need to be recompiled with new version of "generator-helper", but since i'v done that workaround it is no more breaking change...

regarding `prisma-client-js`, as far as i understand how it works - it will recompile entire `client` in node_modules folder so as soon as user tries to "prisma generate" his old version of @prisma/client will be changed to new one... if that's not the case - tell me please and i will change the code to support old @prisma/client.

timsuchanek commented:
Thanks for the explanation @muxahuk!
The problem is, that a user can have a new `@prisma/cli`, which is not printing the message anymore, but an old `@prisma/client`, which does not yet print the message.
It's not the most important thing in the world to be honest, as those users already know how to use Prisma Client, but it would be great to make the transition as smooth as possible.

muxahuk committed:
feat(cli,client,generator-helper,sdk): add fallback for prisma-client-js message

added fallback message to the generate command if we are generating prisma-client-js and it does not return "useMessage", is old version of client

muxahuk commented:
Hey, @timsuchanek added fallback for old prisma-client-js

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prisma/prisma?pullRequest=4231) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prisma/prisma?pullRequest=4231) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prisma/prisma?pullRequest=4231) it.</sub>

wmadden commented:
Thank you for your contribution! As part of our effort to reduce noise and better focus on what matters to the community ([more on that here](https://pris.ly/orm-manifesto)), we’re closing inactive PRs. If this is still relevant, feel free to reopen it or submit a new PR with updated changes. We truly appreciate your efforts and contributions to the project!
